### ISSUE TITLE
Rule Change: allow `void` in `array-callback-return` for `Array.prototype.forEach`

### ISSUE THREAD

**Issue by nopeless** [eslint/eslint#17285]
### What rule do you want to change?

array-callback-return

### What change to do you want to make?

Implement suggestions

### How do you think the change should be implemented?

A new option

### Example code

```js
// example of a non-void side-effect function
[1,2,3].forEach(n => cls.computeCache(n))

// will suggest the following
// Insert void
[1,2,3].forEach(n => void cls.computeCache(n))
// Wrap in {}
[1,2,3].forEach(n => {cls.computeCache(n)})
```

### What does the rule currently do for this code?

Rightfully marks it as wrong, as return value of `forEach`'s callback is ignored

However, even adding a void in front does not fix the issue, and there are no IDE suggestions

### What will the rule do after it's changed?

add a new option `allowVoid`. It will share the same logic that I have implemented for detecting `void` and suggesting fixes in #17282 

If #17282 passes, then it is logical that this rule has consistent options and behavior.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

I think I saw an issue related to `forEach` and `void`, but I can't seem to find it for now

**Comment by sam3k** [comment]
Hey @nopeless, I am not too familiar with this lint rule and I'm trying to repro in the [playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgYXJyYXktY2FsbGJhY2stcmV0dXJuOiBcImVycm9yXCIqL1xuXG52YXIgaW5kZXhNYXAgPSBteUFycmF5LnJlZHVjZShmdW5jdGlvbihtZW1vLCBpdGVtLCBpbmRleCkge1xuICAgIG1lbW9baXRlbV0gPSBpbmRleDtcbn0sIHt9KTtcblxudmFyIGZvbyA9IEFycmF5LmZyb20obm9kZXMsIGZ1bmN0aW9uKG5vZGUpIHtcbiAgICBpZiAobm9kZS50YWdOYW1lID09PSBcIkRJVlwiKSB7XG4gICAgICAgIHJldHVybiB0cnVlO1xuICAgIH1cbn0pO1xuXG52YXIgYmFyID0gZm9vLmZpbHRlcihmdW5jdGlvbih4KSB7XG4gICAgaWYgKHgpIHtcbiAgICAgICAgcmV0dXJuIHRydWU7XG4gICAgfSBlbHNlIHtcbiAgICAgICAgcmV0dXJuO1xuICAgIH1cbn0pO1xuXG5cblxuLy8gZXhhbXBsZSBvZiBhIG5vbi12b2lkIHNpZGUtZWZmZWN0IGZ1bmN0aW9uXG5bMSwyLDNdLmZvckVhY2gobiA9PiBjbHMuY29tcHV0ZUNhY2hlKG4pKVxuXG4vLyB3aWxsIHN1Z2dlc3QgdGhlIGZvbGxvd2luZ1xuLy8gSW5zZXJ0IHZvaWRcblsxLDIsM10uZm9yRWFjaChuID0+IHZvaWQgY2xzLmNvbXB1dGVDYWNoZShuKSlcblxuLy8gV3JhcCBpbiB7fVxuWzEsMiwzXS5mb3JFYWNoKG4gPT4ge2Nscy5jb21wdXRlQ2FjaGUobil9KVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJhcnJheS1jYWxsYmFjay1yZXR1cm4iOlsiZXJyb3IiXX0sImVudiI6eyJlczYiOnRydWV9fX0=) but do not see it marking any of the examples you've provided as wrong. Am I missing something? Could you update the playground example?

**Comment by nopeless** [comment]
[repro](https://eslint.org/play/#eyJ0ZXh0IjoiXG52YXIgaW5kZXhNYXAgPSBteUFycmF5LnJlZHVjZShmdW5jdGlvbihtZW1vLCBpdGVtLCBpbmRleCkge1xuICAgIG1lbW9baXRlbV0gPSBpbmRleDtcbn0sIHt9KTtcblxuQXJyYXkuZnJvbShub2RlcywgZnVuY3Rpb24obm9kZSkge1xuICAgIGlmIChub2RlLnRhZ05hbWUgPT09IFwiRElWXCIpIHtcbiAgICAgICAgcmV0dXJuIHRydWU7XG4gICAgfVxufSk7XG5cblxuW10ubWFwKGZ1bmN0aW9uKG5vZGUpIHtcbiAgICBpZiAobm9kZS50YWdOYW1lID09PSBcIkRJVlwiKSB7XG4gICAgICAgIHJldHVybiB0cnVlO1xuICAgIH1cbiAgICAvLyBpbXBsaWNpdCByZXR1cm5cbiAgICByZXR1cm47XG59KTtcblxudmFyIGJhciA9IGZvby5maWx0ZXIoZnVuY3Rpb24oeCkge1xuICAgIGlmICh4KSB7XG4gICAgICAgIHJldHVybiB0cnVlO1xuICAgIH0gZWxzZSB7XG4gICAgICAgIHJldHVybjtcbiAgICB9XG59KTtcblxuXG5cbi8vIGV4YW1wbGUgb2YgYSBub24tdm9pZCBzaWRlLWVmZmVjdCBmdW5jdGlvblxuWzEsMiwzXS5mb3JFYWNoKG4gPT4gY2xzLmNvbXB1dGVDYWNoZShuKSlcblxuLy8gd2lsbCBzdWdnZXN0IHRoZSBmb2xsb3dpbmdcbi8vIEluc2VydCB2b2lkXG5bMSwyLDNdLmZvckVhY2gobiA9PiB2b2lkIGNscy5jb21wdXRlQ2FjaGUobikpXG5cbi8vIFdyYXAgaW4ge31cblsxLDIsM10uZm9yRWFjaChuID0+IHtjbHMuY29tcHV0ZUNhY2hlKG4pfSlcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiYXJyYXktY2FsbGJhY2stcmV0dXJuIjpbImVycm9yIix7ImNoZWNrRm9yRWFjaCI6dHJ1ZSwiYWxsb3dJbXBsaWNpdCI6dHJ1ZX1dfSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==)

@sam3k 

Sorry, forgot to mention that this only happens when check foreach is enabled

**Comment by mdjermanovic** [comment]
Makes sense to me 👍

**Comment by nzakas** [comment]
@nopeless we need it on a separate pull request so we can evaluate each separately.

**Comment by Gautam-Arora24** [comment]
Would love to work on this issue.

**Comment by sam3k** [comment]
@nopeless will you be able to create a separate PR?

**Comment by nopeless** [comment]
> @nopeless will you be able to create a separate PR?

Yes, but I am planning to do it after #17282 is passed since it would be easier that way

**Comment by Tanujkanti4441** [comment]
@snitin315 are you working on this currently?

**Comment by snitin315** [comment]
@Tanujkanti4441 feel free to work on this.

**Comment by mdjermanovic** [comment]
PR https://github.com/eslint/eslint/pull/17564 that implements the new option `allowVoid` has been merged. The remaining task is to implement suggestions.

**Comment by mdjermanovic** [comment]
Fixed by https://github.com/eslint/eslint/pull/17564 and https://github.com/eslint/eslint/pull/17590.
